File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/sensors/bno086/Shtp.java
package ai.depz.sensor.sensors.bno086;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * SHTP framing layer for the BNO086 (contracts/05_SENSOR_BNO086.md §3).
 *
 * <p>Header: length u16 LE (bits 14:0 = cargo length incl. the 4-byte header;
 * bit 15 = continuation), channel u8, seq u8. A first fragment advertises the
 * TOTAL cargo length; continuations carry the remaining length with bit 15 set.
 * TX seq counters are per channel. Mirror of the TS/Python references.
 */
public final class Shtp {
    private Shtp() {}

    public static final int SHTP_HEADER_SIZE = 4;
    public static final int LENGTH_MASK = 0x7fff;
    public static final int CONTINUATION_BIT = 0x8000;
    public static final int NUM_CHANNELS = 6;
    /** Host->sensor frames must fit one MCU transmit slot (ERRATA E2). */
    public static final int MAX_TX_FRAME = 64;

    public record ShtpHeader(int length, int channel, int seq, boolean continuation) {}

    public static byte[] packShtpHeader(ShtpHeader hdr) {
        int word = (hdr.length() & LENGTH_MASK) | (hdr.continuation() ? CONTINUATION_BIT : 0);
        return new byte[] {
            (byte) word, (byte) (word >>> 8), (byte) hdr.channel(), (byte) hdr.seq()
        };
    }

    public static ShtpHeader unpackShtpHeader(byte[] data) {
        int word = (data[0] & 0xFF) | ((data[1] & 0xFF) << 8);
        return new ShtpHeader(
                word & LENGTH_MASK, data[2] & 0xFF, data[3] & 0xFF, (word & CONTINUATION_BIT) != 0);
    }

    /** One reassembled cargo: {@code payload} excludes all SHTP headers. */
    public record ShtpCargo(int channel, int seq, byte[] payload) {}

    /** Single-fragment frame: length = header + payload. */
    public static byte[] buildFrame(int channel, byte[] payload, int seq) {
        byte[] hdr =
                packShtpHeader(new ShtpHeader(SHTP_HEADER_SIZE + payload.length, channel, seq & 0xff, false));
        byte[] out = new byte[hdr.length + payload.length];
        System.arraycopy(hdr, 0, out, 0, hdr.length);
        System.arraycopy(payload, 0, out, hdr.length, payload.length);
        return out;
    }

    private static final class ChannelRx {
        final ByteArrayOutputStream chunks = new ByteArrayOutputStream();
        int received = 0;
        int expected = 0;
        int seq = 0;

        void clear() {
            chunks.reset();
            received = 0;
            expected = 0;
        }
    }

    /** Per-channel TX sequence counters + RX cargo reassembly. */
    public static final class ShtpLayer {
        /** Incomplete cargos thrown away. */
        public int discarded = 0;

        private final int[] txSeqs = new int[NUM_CHANNELS];
        private final ChannelRx[] rx = new ChannelRx[NUM_CHANNELS];

        public ShtpLayer() {
            for (int i = 0; i < NUM_CHANNELS; i++) {
                rx[i] = new ChannelRx();
            }
        }

        /** Build a single-fragment frame, consuming the channel's TX seq. */
        public byte[] nextFrame(int channel, byte[] payload) {
            if (SHTP_HEADER_SIZE + payload.length > MAX_TX_FRAME) {
                throw new IllegalArgumentException(
                        "TX cargo " + payload.length + "B exceeds the " + MAX_TX_FRAME + "B MCU slot");
            }
            int seq = txSeqs[channel];
            txSeqs[channel] = (seq + 1) & 0xff;
            return buildFrame(channel, payload, seq);
        }

        public int txSeq(int channel) {
            return txSeqs[channel];
        }

        /** Consume one inbound frame; return the cargo when complete, else null. */
        public ShtpCargo feed(byte[] frame) {
            if (frame.length < SHTP_HEADER_SIZE) {
                return null;
            }
            ShtpHeader hdr = unpackShtpHeader(frame);
            if (hdr.channel() >= NUM_CHANNELS || hdr.length() < SHTP_HEADER_SIZE) {
                return null; // empty/padding header or junk
            }
            int chunkLen = frame.length - SHTP_HEADER_SIZE;
            ChannelRx r = rx[hdr.channel()];
            if (!hdr.continuation()) {
                if (r.expected != 0 && r.received != 0) {
                    discarded += 1;
                }
                r.clear();
                r.chunks.write(frame, SHTP_HEADER_SIZE, chunkLen);
                r.received = chunkLen;
                r.expected = hdr.length() - SHTP_HEADER_SIZE;
                r.seq = hdr.seq();
            } else {
                if (r.expected == 0) {
                    discarded += 1;
                    return null;
                }
                r.chunks.write(frame, SHTP_HEADER_SIZE, chunkLen);
                r.received += chunkLen;
            }
            if (r.received < r.expected) {
                return null;
            }
            if (r.received > r.expected) {
                discarded += 1;
                r.clear();
                return null;
            }
            ShtpCargo cargo = new ShtpCargo(hdr.channel(), r.seq, r.chunks.toByteArray());
            r.clear();
            return cargo;
        }

        /** Forget all TX seq counters and partial cargos (sensor reset). */
        public void reset() {
            for (int i = 0; i < NUM_CHANNELS; i++) {
                txSeqs[i] = 0;
                rx[i].clear();
                rx[i].seq = 0;
            }
            discarded = 0;
        }
    }

    /** Convenience for tests/consumers: reassemble a list of frames. */
    public static List<ShtpCargo> reassemble(List<byte[]> frames, int[] discardedOut) {
        ShtpLayer layer = new ShtpLayer();
        List<ShtpCargo> cargos = new ArrayList<>();
        for (byte[] f : frames) {
            ShtpCargo c = layer.feed(f);
            if (c != null) {
                cargos.add(c);
            }
        }
        if (discardedOut != null && discardedOut.length > 0) {
            discardedOut[0] = layer.discarded;
        }
        return cargos;
    }
}

File: packages/depz-sensor-sdk-rust/src/bno086/shtp.rs
//! SHTP framing layer for the BNO086 (contract 05 §3).
//!
//! Pure codec — no I/O. A frame is a 4-byte header plus a cargo fragment:
//!
//! - `length` u16 LE — bits 14:0 cargo length *including* the 4-byte header;
//!   bit 15 set marks a continuation fragment.
//! - `channel` u8, `seq` u8 — per-channel, per-direction free-running counter.
//!
//! A multi-frame cargo's first fragment advertises the TOTAL cargo length;
//! continuations carry the remaining length with bit 15 set (informative).
//!
//! Mirrors the TS/Python `ShtpLayer`.

pub const SHTP_HEADER_SIZE: usize = 4;
pub const LENGTH_MASK: u16 = 0x7fff;
pub const CONTINUATION_BIT: u16 = 0x8000;
pub const NUM_CHANNELS: usize = 6;
/// Host→sensor frames must fit one MCU transmit slot (ERRATA E2).
pub const MAX_TX_FRAME: usize = 64;

/// SHTP channels (contract 05 §3).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum ShtpChannel {
    Command = 0,
    Executable = 1,
    Control = 2,
    InputNormal = 3,
    InputWake = 4,
    GyroRv = 5,
}

/// A decoded SHTP header.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ShtpHeader {
    /// Bits 14:0 — cargo length incl. this 4-byte header.
    pub length: u16,
    pub channel: u8,
    pub seq: u8,
    pub continuation: bool,
}

/// Serialize an SHTP header to 4 bytes.
pub fn pack_shtp_header(hdr: &ShtpHeader) -> [u8; 4] {
    let word = (hdr.length & LENGTH_MASK) | if hdr.continuation { CONTINUATION_BIT } else { 0 };
    let w = word.to_le_bytes();
    [w[0], w[1], hdr.channel, hdr.seq]
}

/// Decode a 4-byte SHTP header (reads the first 4 bytes of `data`).
pub fn unpack_shtp_header(data: &[u8]) -> ShtpHeader {
    let word = u16::from_le_bytes([data[0], data[1]]);
    ShtpHeader {
        length: word & LENGTH_MASK,
        channel: data[2],
        seq: data[3],
        continuation: (word & CONTINUATION_BIT) != 0,
    }
}

/// One reassembled cargo: `payload` excludes all SHTP headers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ShtpCargo {
    pub channel: u8,
    /// seq of the first fragment.
    pub seq: u8,
    pub payload: Vec<u8>,
}

/// Single-fragment frame: length = header + payload.
pub fn build_frame(channel: u8, payload: &[u8], seq: u8) -> Vec<u8> {
    let hdr = pack_shtp_header(&ShtpHeader {
        length: (SHTP_HEADER_SIZE + payload.len()) as u16,
        channel,
        seq,
        continuation: false,
    });
    let mut out = Vec::with_capacity(SHTP_HEADER_SIZE + payload.len());
    out.extend_from_slice(&hdr);
    out.extend_from_slice(payload);
    out
}

#[derive(Debug, Default, Clone)]
struct ChannelRx {
    chunks: Vec<u8>,
    received: usize,
    /// Total cargo payload bytes (headers excluded).
    expected: usize,
    seq: u8,
}

/// Per-channel TX sequence counters + RX cargo reassembly.
#[derive(Debug, Clone)]
pub struct ShtpLayer {
    /// Incomplete cargos thrown away.
    pub discarded: u64,
    tx_seqs: [u8; NUM_CHANNELS],
    rx: Vec<ChannelRx>,
}

impl Default for ShtpLayer {
    fn default() -> Self {
        ShtpLayer {
            discarded: 0,
            tx_seqs: [0; NUM_CHANNELS],
            rx: vec![ChannelRx::default(); NUM_CHANNELS],
        }
    }
}

impl ShtpLayer {
    pub fn new() -> Self {
        ShtpLayer::default()
    }

    /// Build a single-fragment frame, consuming the channel's TX seq.
    pub fn next_frame(&mut self, channel: u8, payload: &[u8]) -> Vec<u8> {
        let ch = channel as usize;
        let seq = self.tx_seqs[ch];
        self.tx_seqs[ch] = seq.wrapping_add(1);
        build_frame(channel, payload, seq)
    }

    /// Current TX seq counter for a channel (next frame's seq).
    pub fn tx_seq(&self, channel: u8) -> u8 {
        self.tx_seqs[channel as usize]
    }

    /// Consume one inbound frame; return the cargo when complete.
    ///
    /// Rules (contract 05 §3): a non-continuation fragment starts a new cargo
    /// (discarding any partial one on that channel); a continuation without a
    /// cargo in progress is dropped; the cargo completes when the accumulated
    /// bytes reach the first fragment's advertised total.
    pub fn feed(&mut self, frame: &[u8]) -> Option<ShtpCargo> {
        if frame.len() < SHTP_HEADER_SIZE {
            return None;
        }
        let hdr = unpack_shtp_header(frame);
        if hdr.channel as usize >= NUM_CHANNELS || (hdr.length as usize) < SHTP_HEADER_SIZE {
            return None; // padding/"no data" header or junk
        }
        let chunk = &frame[SHTP_HEADER_SIZE..];
        let ch = hdr.channel as usize;

        if !hdr.continuation {
            {
                let rx = &self.rx[ch];
                if rx.expected != 0 && rx.received != 0 {
                    self.discarded += 1;
                }
            }
            let rx = &mut self.rx[ch];
            rx.chunks = chunk.to_vec();
            rx.received = chunk.len();
            rx.expected = hdr.length as usize - SHTP_HEADER_SIZE;
            rx.seq = hdr.seq;
        } else {
            if self.rx[ch].expected == 0 {
                self.discarded += 1;
                return None;
            }
            let rx = &mut self.rx[ch];
            rx.chunks.extend_from_slice(chunk);
            rx.received += chunk.len();
        }

        let rx = &mut self.rx[ch];
        if rx.received < rx.expected {
            return None;
        }
        if rx.received > rx.expected {
            self.discarded += 1;
            rx.chunks.clear();
            rx.received = 0;
            rx.expected = 0;
            return None;
        }
        let cargo = ShtpCargo {
            channel: hdr.channel,
            seq: rx.seq,
            payload: std::mem::take(&mut rx.chunks),
        };
        rx.received = 0;
        rx.expected = 0;
        Some(cargo)
    }

    /// Forget all TX seq counters and partial cargos (sensor reset).
    pub fn reset(&mut self) {
        self.tx_seqs = [0; NUM_CHANNELS];
        self.rx = vec![ChannelRx::default(); NUM_CHANNELS];
    }
}
